ref: main exp workflow (run,show,diff,apply) (#2200)
iterative/dvc.org: content/docs/api-reference/make_checkpoint.md

## Description

To track successive steps in a longer <abbr>experiment</abbr>, you can write
your code so it registers checkpoints with DVC during runtime (similar to a

> logger). This function can be called by the code in stages executes by

Reviewer: ```suggestion
logger). This function can be called by the code in stages executed by
```

> `dvc exp run` (see also the `cmd` field of `dvc.yaml`).

Reviewer: Is this supposed to refer to the `checkpoint` subfield of `output`? I'm unclear how `cmd` is related. Does it make sense to include an example `dvc.yaml`?

Author: No, I meant to mention `cmd` but yeah it's not clear, thanks. Updated

> Note that for non-Python code, the alternative is to write a

> > `.dvc/tmp/DVC_CHECKPOINT` signal file.

Author: Another QQ @pmrowla: How exactly does this work? You write that file (empty I suppose) and DVC picks it up and deletes it? How can you write one per process for `run -j`? Thanks

Author: Or one per output

Reviewer 2: Parallel runs with `-j` are run in their own temporary workspaces, so if I have my regular DVC/git workspace at `~/dvc-project`, and then do `exp run -j 2`, we create copies of the project at `~/dvc-project/.dvc/tmp/exp1` and `~/dvc-project/.dvc/tmp/exp2`. These copies are their own independent DVC workspaces (with their own `.dvc/` directories), but they are not full git clones - the temporary workspaces only start with the single experiment commit to reproduce rather than the full git history.

Essentially, the way execution works is that we `cd` into `~/dvc-project/.dvc/tmp/exp1` and then run `dvc repro`. So for generating the checkpoints for experiment one, the full signal path is `~/dvc-project/.dvc/tmp/exp1/.dvc/tmp/DVC_CHECKPOINT` (which is independent of the second experiment's signal path inside `~/dvc-project/.dvc/tmp/exp2/...`)

The user's code (possibly via `dvc.api.make_checkpoint()`) should be generating the signal file by first checking the `$DVC_ROOT` environment variable (so the signal path is `$DVC_ROOT/tmp/DVC_CHECKPOINT`. `DVC_ROOT` points to the location of `.dvc/`. For a temp directory experiment run, it will point to the `.dvc` directory for that specific temporary run, and not the `.dvc` directory for the "main" repo.

Reviewer 2: Also the checkpoint generation & signaling has nothing to do with individual outputs. When the user's code signals to make a checkpoint, we save the state of the entire repo (including any and all checkpoint outputs).

Reviewer 2: And again, this is all implementation detail that the user should not care about, all they need to know is that if they aren't using `dvc.api.make_checkpoint`, they just have to create an empty file at `$DVC_ROOT/tmp/DVC_CHECKPOINT` to do the signaling.

Reviewer 2: I think the signaling process for non-python code does need to be more thoroughly documented somewhere though.
They should be directed to refer to the actual `api.make_checkpoint()` implementation as well, but the way it should work is:

1. User's code does what it needs to and eventually reaches a point where they want to generate a checkpoint commit
2. User checks the `DVC_ROOT` environment variable, and then creates a file `$DVC_ROOT/tmp/DVC_CHECKPOINT` (it can contain anything or be an empty file, DVC only cares that the file exists)
3. User's code should block as long as the file `$DVC_ROOT/tmp/DVC_CHECKPOINT` still exists.
    * Making this a blocking call is important - the user's code should explicitly not do anything that writes to/modifies the repo/workspace state, as DVC will be staging files and creating commits in git during this time
    * The user should also ensure that the file handle used to open the signal file in step 2 has been closed (on windows if the file handle is still open, DVC may not be able to remove the file later, which is the signal for the user's code to continue)
4. When the file `$DVC_ROOT/tmp/DVC_CHECKPOINT` no longer exists (because DVC has removed it), this is the signal that the user's code can resume. Go back to step 1 and repeat checkpoint generation as needed

Author: Interesting @pmrowla! (BTW I ask because I need to know so thanks for the answers). I've put the relevant details in the refs now (see 53f02ab2). Some questions though:

> copies are their own independent DVC workspaces (with their own .dvc/ directories

Including a copy of the cache? (I guess not)

> the checkpoint generation & signaling has nothing to do with individual outputs

Yeah I thought that was strange. But the `checkpoint: true` field is under `outs` in dvc.yaml, apparently. Why not under the stage directly?

Why do you need that field at all? Curious

> the signaling process for non-python code does need to be more thoroughly documented somewhere

- [ ] Agreed

> 2. User checks DVC_ROOT and creates a file $DVC_ROOT/tmp/DVC_CHECKPOINT
> 3. User's code should block as long as the file still exists

Is that what `make_checkpoint()` does internally? (Assumed so for now)

Reviewer 2: > Including a copy of the cache?

The original repo's cache dir is used as the cache for the temporary repo

> Why not under the stage directly?
> Why do you need that field at all?

Checkpoint outs are treated differently than regular outputs, one of difference is that reproduction of a stage and checkout won't fail if a checkpoint out does not exist. So the flag applies to specific outputs rather than an entire stage.

> Is that what make_checkpoint() does internally? 

Yes

In this example we kill the process (with Ctrl + C) after 3 checkpoints (at 0,
100, and 200). The <abbr>cache</abbr> will contain those 3 versions of

> `int.txt`. DVC applies the last checkpoint to the <abbr>workspace</abbr> even
> when more cycles happened before the interrupt:

Reviewer: ```suggestion
`int.txt`. DVC applies the last checkpoint to the <abbr>workspace</abbr>
regardless of how many cycles happened before the interrupt:
```
I don't understand the current sentence, so I'm not sure if this captures the intended meaning.

Author: Changing to "if more cycles happened after that" but this example is kind of a placeholder for now. I'll prob steal/adapt your GS examples for the other docs in a following iteration 😬 Thanks though!
